### PR TITLE
Improve IPv6 support in ports

### PR DESCRIPTION
**Pull request by MaciejPurski**

### CODE REVIEW

File: port/sockets.c [hunk 1]
```diff
@@ -169,6 +169,111 @@ static void inet6_addr_netmask_from_ip6addr(struct in6_addr *dst, const ip6_addr
 	}
 }
 
+static int socket_ioctl6(int sock, unsigned long request, const void *in_data, const void *out_data)
```
Comment (nalajcie) [inline port/sockets.c:172]: remove `const` from `const void *out_data`

**Review by nalajcie — COMMENTED** [review]
Overall seems good, I've pointed out some minor programming issues which might be worth resolving, but they are not a merge blocker

File: port/sockets.c [hunk 2]
```diff
@@ -153,7 +149,7 @@ static int socket_ioctl(int sock, unsigned long request, const void* in_data, vo
 			if (interface == NULL)
 				return -ENXIO;
 
-			ifreq->ifr_ifindex = interface->num;
+			ifreq->ifr_ifindex = netif_get_index(interface);
```
[the remark this answers is not in the document]
Comment (MaciejPurski) [reply port/sockets.c:307]: This change does not affect interface names. The interface `index` and interface `number` are in lwip different notions. Interface `index` is defined in in [RFC 3493](https://tools.ietf.org/html/rfc3493#section-4). On interface index the RFC says:
> These are small positive integers that start at 1.  (Note that 0 is never used for an interface index.)
I encountered this anomaly, when I was calling `connect()` function with IPv6 `sockaddr_in6` structure, where `lwip` expects *interface index* (starting with 1) in `sin6_scope_id` field in order to decide, which interface to use, when dealing with link-local addresses. In `getaddrinfo()` standard function lwip also sets this field to *interface index* as stated in the RFC.

Going back to interface names, lwip assumes they are 2-letters `netif->name` concatenated with `netif->num`, so the names are "lo0", "en1", etc. You can see it in `netif_find()` or `netif_index_to_name()`.

So, to sum up, we should check, wether `if_nametoindex()` is used to somewhere and if the returned value is used in order to create interface names. If yes, then we have a problem and we have to either give this change up and remember in the future, that our implementation does not conform to the RFC or fix all the occurrences of this function.

File: port/sockets.c [hunk 3]
```diff
@@ -737,6 +787,109 @@ static int do_getaddrinfo(const char *name, const char *serv, const struct addri
...
+	memset(&sa, 0, sizeof(sa));
+	sin = (struct sockaddr_in *) &sa;
+	NETIF_FOREACH(netif) {
+		dest->ifa_flags = netif->flags;
+		sin->sin_family = AF_INET;
+		sin->sin_len = sizeof(struct sockaddr_in);
+
+		inet_addr_from_ip4addr(&sin->sin_addr, netif_ip4_addr(netif));
+		memcpy(addrdest, sin, sin->sin_len);
+		dest->ifa_addr = (void *) (addrdest - buf);
```
[the remark this answers is not in the document]
Comment (MaciejPurski) [reply port/sockets.c:962]: Yes, that's how it works. That's, how it is done in `getaddrinfo()`.
I'll change `addrdest` and `buf` types to `char *`. But again, it's a mistake I copied from `getaddrinfo()` ;) So we might need to fix it there too.

File: port/sockets.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (nalajcie) [reply port/sockets.c:307]: Ok, so this change won't check the actual interface names. Thanks for the detailed description.

File: port/sockets.c [hunk 4]
```diff
@@ -136,15 +136,11 @@ static int socket_ioctl(int sock, unsigned long request, const void* in_data, vo
...
-		for (it = netif_list; it != NULL; it = it->next) {
-			if (it->num == ifreq->ifr_ifindex) {
-				strncpy(ifreq->ifr_name, it->name, IFNAMSIZ);
-				return EOK;
-			}
-		}
+		char *res;
 
-		return -ENXIO;
+		res = netif_index_to_name(ifreq->ifr_ifindex, ifreq->ifr_name);
```
[the remark this answers is not in the document]
Comment (MaciejPurski) [reply port/sockets.c:296]: `IFNAMSIZ` and `NETIF_NAMESIZE` are quite problematic. lwip defines `IFNAMSIZ` to be equal to `NETIF_NAMESIZE`, which is equal to 6. libphoenix defnies `IFNAMSIZ` to 16. We use `libphoenix` definition of `ifreq` (look at the `#undef ifreq` hack at te file beggining), so our size of the name buffer is 16. In `netif_index_to_name()` an `lwip_itoa()` function is used, which is safe, given the buffer is at least `NETIF_NAMESIZE` long. If we can trust, that the `ifreq` pointer indeed points to a `struct ifreq` type object, then we are safe. However, if you'd like to make it more explicit, we could add this kind of assertion, if someone decided to change IFNAMESIZ value. I can't think of any other sanity-check in this case.
